### PR TITLE
[bazel] Rewrite binary & test rules 

### PR DESCRIPTION
**Pull request by cfrantz**
1. Create rules for describing the target execution environments for
   OpenTitan code, including "arch" libraries, linker scripts, signing
   keys and test dispatch information.
2. Create a new `opentitan_binary` rule that can build and link C
   programs for a set of execution environments.  Emit providers
   specific to each environment.
3. Createa new `opentitan_test` rule that can dispatch a test on a
   given execution environment.  Like `cc_test`, the opentitan rule
   can compile and link the test program before dispatch.
4. Create an `opentitan_test` macro that, given a list of execution
   environments, creates a rule per exec_env.

# Note
This is still a work in progress.  Draft pull request to collect some early feedback.

### CODE REVIEW

File: rules/opentitan/cc.bzl [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    return lout.executable, mapfile
+
+def _as_group_info(name, items):
+    """Prepare a dict of files for OutputGroupInfo.
+
+    Args:
+      name: A prefix for each dictionary key.
+      items: A dict str:File to prepare.
+    Returns:
+      dict
```
Comment (pamaury) [inline rules/opentitan/cc.bzl:113]: Can you document a bit more the names of the groups generated, or add an example?

File: rules/opentitan/cc.bzl [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        doc = "Cleanup script for the disassembly",
+    ),
+}
+
+opentitan_binary = rv_rule(
+    implementation = _opentitan_binary,
+    attrs = dict(common_binary_attrs.items() + {
+        "exec_env": attr.label_list(
+            providers = [ExecEnvInfo],
+            doc = "List of exeuction environments for this target.",
```
Comment (pamaury) [inline rules/opentitan/cc.bzl:259]: *execution

File: rules/opentitan/cc.bzl [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        ),
+        "_cc_toolchain": attr.label(default = Label("@bazel_tools//tools/cpp:current_cc_toolchain")),
+    }.items()),
+    fragments = ["cpp"],
+    toolchains = ["@rules_cc//cc:toolchain_type"],
+)
+
+def _opentitan_test(ctx):
+    exec_env = ctx.attr.exec_env[ExecEnvInfo]
+    if len(ctx.attr.srcs) == 1 and len(ctx.attr.deps) == 0 and exec_env.get_provider(ctx.attr.srcs[0]):
```
Comment (pamaury) [inline rules/opentitan/cc.bzl:269]: I assume this is the case where we provide `_opentitan_test` an already built binary, versus a source file?

File: rules/opentitan/cc.bzl [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        p = exec_env.get_provider(ctx.attr.srcs[0])
+    else:
+        name = _binary_name(ctx, exec_env)
```
Comment (pamaury) [inline rules/opentitan/cc.bzl:272]: Why does this branch reimplements the inner loop of `_opentitan_binary` versus just called `_opentitan_binary`?

File: rules/opentitan/cc.bzl (hunk 1 above)
Comment (cfrantz) [reply rules/opentitan/cc.bzl:116]: Done

File: rules/opentitan/cc.bzl (hunk 2 above)
Comment (cfrantz) [reply rules/opentitan/cc.bzl:284]: Fixed.

File: rules/opentitan/cc.bzl (hunk 3 above)
Comment (cfrantz) [reply rules/opentitan/cc.bzl:294]: Yes.  Added a comment.

File: rules/opentitan/cc.bzl (hunk 4 above)
Comment (cfrantz) [reply rules/opentitan/cc.bzl:297]: `_opentitan_binary` expects to deal with a list of `exec_env` whereas the test rule expects to have only one `exec_env`.  The binary rule also produces a bunch of output providers that are not relevant to the test rule.

File: rules/opentitan.bzl [hunk 5]
```diff
@@ -625,6 +625,7 @@ def _gen_sim_dv_logs_db_impl(ctx):
         outputs.append(logs_db)
         outputs.append(rodata)
 
+        print("extool:", ctx.executable._tool)
```
Comment (timothytrippel) [inline rules/opentitan.bzl:628]: do you need this?

File: rules/opentitan/exec_env.bzl [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    """Common attributes for rules creating ExecEnvInfo providers."""
+    return {
+        "base": attr.label(
+            default = kwargs.get("base"),
+            providers = [ExecEnvInfo],
+            doc = "Base execution environment used to initialize this environment",
+        ),
+        "design": attr.string(
+            default = kwargs.get("design", ""),
+            doc = "Hardware design name (e.g. `earlgrey`)",
```
Comment (timothytrippel) [inline rules/opentitan/exec_env.bzl:90]:
```suggestion
            doc = "Top-level hardware design name (e.g. `earlgrey`)",
```
?

File: rules/opentitan/util.bzl [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    items = item.split(".")
+    item = items[-1]
+    if item in overrides:
+        return overrides.get(item)
+    for i in items:
+        obj = getattr(obj, i)
+    return obj
+
+def get_fallback(obj, item, fallback):
+    """Get an item from obj and fall back to `fallback if falsy.
```
Comment (timothytrippel) [inline rules/opentitan/util.bzl:23]: nit:
```suggestion
    """Get an item from obj and fall back to `fallback` if falsy.
```

**Review by timothytrippel — APPROVED** [review]
wow this is much cleaner than what we have!

File: rules/signing.bzl [hunk 8]
```diff
@@ -339,36 +340,48 @@ offline_signature_attach = rule(
     },
 )
 
-def _sign_bin_impl(ctx):
-    rsa_key = key_from_dict(ctx.attr.rsa_key, "rsa_key")
-    spx_key = key_from_dict(ctx.attr.spx_key, "spx_key")
+def sign_binary(ctx, **kwargs):
```
Comment (pamaury) [inline rules/signing.bzl:343]: Maybe document the content of the returned dictionary?

**Comment by pamaury** [comment]
I have gone through the change and while I can't claim to understand every single detail, I think it makes sense overall. It should be a nice improvement over the current state. One thing is that is a bit hard to follow is that a lot of information is passed through `ctx` and overidden by arguments, and this is not necessarily documented for each function. One such example is `sign_binary`. I can see why it is done like that but I also see how this could cause subtle/hard to understand bugs where some fields are ignored or taken into account in surprising ways further down the line when the code is modified. I am not sure how/if this can improved, except maybe by some documentation.

File: rules/opentitan/exec_env.bzl (hunk 6 above)
Comment (cfrantz) [reply rules/opentitan/exec_env.bzl:90]: Fixed

File: rules/opentitan/util.bzl (hunk 7 above)
Comment (cfrantz) [reply rules/opentitan/util.bzl:23]: Fixed

File: rules/signing.bzl (hunk 8 above)
Comment (cfrantz) [reply rules/signing.bzl:343]: Done.

**Review by pamaury — APPROVED** [review]
